arch/riscv/core/cpu_idle.c

```diff
@@ -10,13 +10,15 @@
 void __weak arch_cpu_idle(void)
 {
 	sys_trace_idle();
-	irq_unlock(MSTATUS_IEN);
+	arch_irq_lock();
```

Comment:
Hi, thanks for the PR. I’ve got a question: it looks like the `arch_irq_lock` call that you’re introducing here duplicates the one present here: https://github.com/zephyrproject-rtos/zephyr/blob/main/kernel/idle.c#L65-L90, i.e. the one just before we call the `k_cpu_idle`. In general, we seem to duplicate here the functionality that already exists in the common code. I suppose we don’t need to disable the irqs here again, just move the irq_unlock after WFI.

Reply:
[Docs state that](https://docs.zephyrproject.org/latest/hardware/porting/arch.html#cpu-idling-power-management) `arch_cpu_idle` can be called from idle thread (one you mentioned) and

>In a single-threaded system, in the only thread when the thread is not used for doing real work after initialization

Is it safe to assume that the second case will also have interrupts disabled when entering `arch_cpu_idle`?

Reply:
Interestingly, the function-specific documentation states something different: https://docs.zephyrproject.org/latest/hardware/porting/arch.html#c.arch_cpu_idle.

> This function will be called by the kernel idle loop or possibly within an implementation of z_pm_save_idle in the kernel when the ‘_pm_save_flag’ variable is non-zero.

If that’s the case, then I’d say it’s pretty safe to assume that proper conditions are met when calling the `arch_cpu_idle` function. We’re not expecting this function to be used directly in e.g. user applications. Plus, other archs seem to also follow this assumption.

Reply:
But then in the link you've provided, description for `k_cpu_atomic_idle()` is:
>Similar to `k_cpu_idle()` but must be called with interrupts locked.

Which suggests `k_cpu_idle()` could be called with interrupts unlocked.
Looks like documentation is a bit vague here.
I'm fine with either solution though - I'll remove locking IRQs in both functions.